[ryenl]iP

ryenl opened the pull request:
# King Chatbot Features 🤖

The **King Chatbot** is a simple yet powerful task management assistant. It handles user commands to manage a to-do list, deadlines, and events. Below are the general features of the chatbot:

## Key Features:

- **Task Management**:
  - Add, delete, mark, and unmark tasks.
  - Handle three types of tasks:
    - **ToDo**: Simple tasks without a deadline.
    - **Deadline**: Tasks with a specific due date.
    - **Event**: Tasks with a scheduled duration.
  
- **Storage**:
  - Automatically saves tasks to a file after each update.
  - Loads saved tasks when the chatbot starts.

- **User-Friendly Interaction**:
  - Greets the user upon startup.
  - Exits gracefully when the user types `bye`.

## How to Use:

1. Start the chatbot.
2. Enter commands to manage your tasks:
   - `todo [description]`
   - `deadline [description] /by [date]`
   - `event [description] /from [date] /to [date]`
3. Use the `list` command to view all tasks.
4. Type `bye` to exit the chatbot.

### Example Code Snippet:
```java
public String getGreeting() {
    return "Welcome to the King chatbot!";
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

ryenl committed:
Level-0

ryenl committed:
Level-1

ryenl committed:
Level-2

ryenl committed:
Level-3

ryenl committed:
Level-4

ryenl committed:
Level-5

ryenl committed:
Level-5

ryenl committed:
Level-6

ryenl committed:
A-TextUiTesting (WIP*)

ryenl committed:
Add save feature to King

ryenl committed:
Add date and time recognition

ryenl committed:
Add more OOP

ryenl committed:
Add packages

ryenl committed:
Merge branch 'add-gradle-support'

ryenl committed:
Add gradle support

ryenl committed:
Add JUnit test for tasks

ryenl committed:
Add javadocs

ryenl committed:
Fix coding style and add more javadocs

ryenl committed:
Add find function

FionaQY commented on src/main/java/king/TaskList.java:
```diff
@@ -0,0 +1,109 @@
+     *
+     * @param index the index of the task to be removed
+     */
+    public void remove(int index) {
```
Could throw a KingException instead if index is invalid

FionaQY commented on src/main/java/king/TaskList.java:
```diff
@@ -0,0 +1,109 @@
+     * @param index the index of the task to be retrieved
+     * @return the task at the specified index, or null if the index is out of bounds
+     */
+    public Task getTask(int index) {
```
Could throw a KingException instead if index is invalid

FionaQY commented on src/main/java/king/commands/DeadlineCommand.java:
```diff
@@ -0,0 +1,66 @@
+     */
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage) throws KingException {
+        String[] parts = this.input.split(" /by ");
+        if (parts.length == 2) {
+            String description = parts[0].trim();
+            String by = parts[1].trim();
```
Could move this part into the Parser instead

FionaQY commented on src/main/java/king/commands/Command.java:
```diff
@@ -0,0 +1,30 @@
+     *
+     * @return {@code true} if the command is an exit command, {@code false} otherwise.
+     */
+    public abstract boolean isExit();
```
Given isExit returns false for all commands except ExitCommand, you could also have this command return false by default and override it in ExitCommand

FionaQY reviewed (commented):
LGTM. Just some nits to fix.

LOLIPOP-INTELLIGENCE commented on src/main/java/king/TaskList.java:
```diff
@@ -0,0 +1,109 @@
+     * @param taskNumber the task number to be marked as done
+     * @throws KingException if the task number is invalid
+     */
+    public void markTaskAsDone(int taskNumber) throws KingException {
```
The exception name "KingException" should be replaced with a more standard and descriptive name, such as "InvalidTaskNumberException".

LOLIPOP-INTELLIGENCE commented on src/main/java/king/task/Deadline.java:
```diff
@@ -0,0 +1,57 @@
+ * Represents a task with a deadline.
+ */
+public class Deadline extends Task {
+    protected LocalDateTime by;
```
The variable name "by" is too short and not sufficiently descriptive. It should be renamed to something more explanatory, such as "deadlineDateTime" or "dueDateTime".

LOLIPOP-INTELLIGENCE commented on src/main/java/king/task/Deadline.java:
```diff
@@ -0,0 +1,57 @@
+     * @return A string representing the task in a file format.
+     */
+    @Override
+    public String toFileString() {
```
The method name "toFileString()" could be more descriptive. Consider renaming it to something like "serializeToString()" or "convertToFileFormat()".

ryenl committed:
Add gradle checkstyle plugin

ryenl committed:
Add GUI using JavaFX

ryenl committed:
Edit the code based on Checkstyle

ryenl committed:
Add welcome greeting and exit on "bye" keyword

King did not greet the user on startup, and the chatbot exited without
displaying the goodbye message after the "bye" command was given.

This commit adds a welcome greeting displayed when the chatbot starts,
and ensures that the chatbot  exits when "bye" is inputted by the user.

- Inject a greeting using Ui's showWelcome method during startup
- Exit after "bye" keyword is inputted

ryenl committed:
Add assertions to validate indexes and non-null file path

- Current Situation: Previously, there were no checks to ensure that indexes used in delete and mark commands were valid. Additionally, there was no validation to ensure that the file path for task storage was not null.
- Why It Needs to Change: Without proper validation, invalid indexes could cause runtime errors or unintended behavior, and a null file path could lead to errors in file operations.
- What Is Being Done:
  - Added assertions in the `remove` method to ensure the index is within the valid range before attempting to delete a task.
  - Added assertions in the `markTaskAsDone` and `markTaskAsUndone` methods to ensure the task number is within the valid range before marking tasks.
  - Included an assertion to check that the file path is not null before proceeding with file operations.
- Why: These changes improve the robustness of the code by ensuring that invalid states are caught early, thus preventing potential errors and enhancing overall code reliability.

ryenl committed:
Improve code quality

ryenl committed:
Improve code quality

ryenl committed:
Use streams for find command.

ryenl committed:
Merge pull request #1 from ryenl/branch-A-Assertions

ryenl committed:
Merge branch 'master' into branch-A-CodeQuality

ryenl committed:
Merge pull request #2 from ryenl/branch-A-CodeQuality

Improve code quality

ryenl committed:
Merge pull request #3 from ryenl/branch-A-Streams

Use streams for find command.

ryenl committed:
Add reminders for tasks

ryenl committed:
Add resizable GUI

ryenl committed:
Fix king image and user image orientation

ryenl committed:
Fix king image and user image orientation

ryenl committed:
Add Ui.png

ryenl committed:
Update README to correct event format

ryenl committed:
Fix storing tasks when directory does not exist

ryenl committed:
Add FindCommand and TaskList JUnit tests

ryenl committed:
Fix bug where start date can be set after end date for Event task

ryenl committed:
Update user guide to include GUI
